airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/BigQueryDestination.kt

```diff
@@ -83,33 +98,221 @@ class BigQueryDestination : BaseConnector(), Destination {
+            }
+
             if (UploadingMethod.GCS == uploadingMethod) {
-                val status = checkGcsPermission(config)
+
+                val status = checkGcsAccessPermission(config)
                 if (status!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
                     return status
                 }
+
+                //Copy a file to confirm copy permissions are working
+                val bigQueryStatus = checkBigQueryCopyPermission(config);
+
+                if (bigQueryStatus!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
+                    return bigQueryStatus
+                }
+
             }
 
-            val result = BigQueryUtils.executeQuery(bigquery, queryConfig)
-            if (result.getLeft() != null) {
-                return AirbyteConnectionStatus()
-                    .withStatus(AirbyteConnectionStatus.Status.SUCCEEDED)
+            return AirbyteConnectionStatus().withStatus(AirbyteConnectionStatus.Status.SUCCEEDED)
+
+        } catch (e: Exception) {
+            log.error(e) { "Check failed." }
+            throw ConfigErrorException((if (e.message != null) e.message else e.toString())!!)
+        }
+
+    }
+
+    /**
+     * This method performs a copy operation to copy data into a temporary table on BigQuery
+     * to check if the existing permissions are sufficient to copy data.
+     * If the permissions are not sufficient, then an exception is thrown with a message
+     * showing the missing permission
+     */
+    private fun checkBigQueryCopyPermission(config: JsonNode): AirbyteConnectionStatus? {
+
+        //TODO: Need to add a step in this method to first check permissions using testIamPermissions before trying the actual copying of data
+
+        val datasetLocation = BigQueryUtils.getDatasetLocation(config)
+        val bigquery = getBigQuery(config)
+
+        val gcsNameTransformer = GcsNameTransformer()
+        val gcsConfig = BigQueryUtils.getGcsCsvDestinationConfig(config)
+        val keepStagingFiles = BigQueryUtils.isKeepFilesInGcs(config)
+        val gcsOperations =
+            GcsStorageOperations(gcsNameTransformer, gcsConfig.getS3Client(), gcsConfig)
+
+        val projectId = config[bqConstants.CONFIG_PROJECT_ID].asText()
+
+        val destinationHandler = BigQueryDestinationHandler(bigquery, datasetLocation)
+
+        val sqlGenerator = BigQuerySqlGenerator(projectId, datasetLocation)
+
+        val defaultDataset = BigQueryUtils.getDatasetId(config)
+
+
+        val finalTableName =
+            "_airbyte_bigquery_connection_test_" +
+                UUID.randomUUID().toString().replace("-".toRegex(), "")
+
+        val rawDatasetOverride: String =
+            if (getRawNamespaceOverride(bqConstants.RAW_DATA_DATASET).isPresent) {
+                getRawNamespaceOverride(bqConstants.RAW_DATA_DATASET).get()
             } else {
-                throw ConfigErrorException(result.right)
```

Comment:
The entire function is new, however, github UI is incorrectly showing this line as being deleted